### PR TITLE
implement roles

### PR DESCRIPTION
**Pull request by Roma1990Melnichuk**

### CODE REVIEW

**Comment by Roma1990Melnichuk** [comment]
Help please - I don't understand what's wrong with the HW

File: src/main/java/com/bookstore/entity/Role.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+@Getter
+@Setter
+@Entity
+@Table(name = "roles")
+public class Role {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Column(columnDefinition = "varchar", nullable = false)
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/entity/Role.java]:
```suggestion
    @Column(columnDefinition = "varchar", nullable = false, unique = true)
```

File: src/main/java/com/bookstore/entity/Role.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    @Enumerated(EnumType.STRING)
+    private RoleName name;
+
+    public enum RoleName {
+        ROLE_USER,
+        ROLE_ADMIN,
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/entity/Role.java]:
```suggestion
        ROLE_ADMIN
```

File: src/main/java/com/bookstore/entity/User.java [hunk 3]
```diff
@@ -31,4 +40,44 @@ public class User {
     private String lastName;
 
     private String shippingAddress;
+
+    @ManyToMany(fetch = FetchType.EAGER)
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/entity/User.java]: Don't use Eager

```suggestion
    @ManyToMany
```

File: src/main/java/com/bookstore/entity/User.java [hunk 4]
```diff
@@ -31,4 +40,44 @@ public class User {
...
+    @JoinTable(
+            name = "users_roles",
+            joinColumns = @JoinColumn(name = "user_id"),
+            inverseJoinColumns = @JoinColumn(name = "role_id")
+    )
+    private Set<Role> roles;
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/entity/User.java]:
```suggestion
    private Set<Role> roles = new HashSet<>();
```

File: src/main/java/com/bookstore/entity/User.java [hunk 5]
```diff
@@ -31,4 +40,44 @@ public class User {
...
+
+    @Override
+    public Collection<? extends GrantedAuthority> getAuthorities() {
+        return roles.stream()
+                .map(r -> new SimpleGrantedAuthority(r.getName().name()))
+                .toList();
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/entity/User.java]:
```suggestion
        return roles;
```

File: src/main/java/com/bookstore/security/CustomUserDetailsService.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+@Service
+@RequiredArgsConstructor
+public class CustomUserDetailsService implements UserDetailsService {
+    private final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String email)
+            throws UsernameNotFoundException {
+        return userRepository.findByEmail(email)
+                .orElseThrow(() -> new UsernameNotFoundException("Can`t find user by email"));
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/security/CustomUserDetailsService.java]:
```suggestion
            {
        return userRepository.findByEmail(email)
                .orElseThrow(() -> new EntityNotFoundException("Can`t find user by email"));
```

File: src/main/java/com/bookstore/security/UserDetails.java [hunk 7]
```diff
@@ -0,0 +1,20 @@
+package com.bookstore.security;
+
+import java.io.Serializable;
+import java.util.Collection;
+import org.springframework.security.core.GrantedAuthority;
+
+public interface UserDetails extends Serializable {
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/security/UserDetails.java]: remove

File: src/main/resources/db/changelog/changes/04-create-roles-table.yaml [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+                  name: id
+                  type: BIGINT
+                  autoIncrement: true
+                  constraints:
+                    primaryKey: true
+              - column:
+                  name: name
+                  type: varchar(255)
+                  constraints:
+                    nullable: false
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/04-create-roles-table.yaml:19]: add unqiue constraint

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/com/bookstore/entity/User.java (hunk 4 above)
Comment (Roma1990Melnichuk) [reply src/main/java/com/bookstore/entity/User.java]: Incompatible types. Found: 'java.util.Set<com.bookstore.entity.Role>', required: 'java.util.Collection<? extends org.springframework.security.core.GrantedAuthority> 
I couldn't fix this error

File: src/main/java/com/bookstore/config/SecurityConfig.java [hunk 9]
```diff
@@ -24,9 +26,17 @@ public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Excepti
                 .cors(AbstractHttpConfigurer::disable)
                 .csrf(AbstractHttpConfigurer::disable)
                 .authorizeHttpRequests(auth -> auth
-                        .requestMatchers("/auth/**", "/error", "/swagger-ui/**").permitAll()
+                        .requestMatchers("/swagger-ui/**",
+                                "/v3/api-docs/**", "/auth/registration").permitAll()
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/config/SecurityConfig.java]: why do you change this `"/auth/**"` to `"/auth/registration"`

File: src/main/java/com/bookstore/controller/BookController.java [hunk 10]
```diff
@@ -29,13 +30,16 @@
 public class BookController {
     private final BookService bookService;
 
+    @PreAuthorize("hasRole('ROLE_USER')")
     @Operation(summary = "Get all books",
             description = "Retrieve a list of all books with pagination and sorting options")
+
     @GetMapping
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/controller/BookController.java:37]:
```suggestion
    @GetMapping
    @PreAuthorize("hasAnyRole('ROLE_ADMIN', 'ROLE_USER')")
```

File: src/main/java/com/bookstore/entity/Role.java [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+import jakarta.persistence.Id;
+import jakarta.persistence.Table;
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+@Entity
+@Table(name = "roles")
+public class Role {
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/entity/Role.java]:
```suggestion
public class Role implements GrantedAuthority {
```

File: src/main/java/com/bookstore/entity/Role.java [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Column(columnDefinition = "varchar", nullable = false, unique = true)
+    @Enumerated(EnumType.STRING)
+    private RoleName name;
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/entity/Role.java:26]: Override method `getAuthority`

File: src/main/java/com/bookstore/entity/User.java [hunk 13]
```diff
@@ -31,4 +40,44 @@ public class User {
...
+            joinColumns = @JoinColumn(name = "user_id"),
+            inverseJoinColumns = @JoinColumn(name = "role_id")
+    )
+    private Set<Role> roles = new HashSet<>();
+
+    @Override
+    public Collection<? extends GrantedAuthority> getAuthorities() {
+        return roles.stream()
+                .map(role -> new SimpleGrantedAuthority(role.getName().name()))
+                .toList();
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/entity/User.java]:
```suggestion
        return roles;
```

File: src/main/java/com/bookstore/entity/User.java [hunk 14]
```diff
@@ -31,4 +40,44 @@ public class User {
...
+
+    @Override
+    public boolean isAccountNonLocked() {
+        return true;
+    }
+
+    @Override
+    public boolean isCredentialsNonExpired() {
+        return true;
+    }
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/entity/User.java:77]: See task description

Think about how we should add columns like is_deleted. Should we add them directly to existing scripts or there is another better way?

File: src/main/java/com/bookstore/security/UserDetails.java (hunk 7 above)
Comment (Elena-Bruyako) [reply src/main/java/com/bookstore/security/UserDetails.java]: remove interfase, you should use Spring Security `UserDetails`

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/com/bookstore/entity/Role.java [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+public class Role implements GrantedAuthority {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @Column(columnDefinition = "varchar", nullable = false, unique = true)
+    @Enumerated(EnumType.STRING)
+    private RoleName name;
+
+    @Column(name = "is_deleted", nullable = false)
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/entity/Role.java]:
```suggestion
    @Column(nullable = false)
```

File: src/main/java/com/bookstore/entity/Role.java [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+    private boolean isDeleted = false;
+
+    @Override
+    public String getAuthority() {
+        return "ROLE_" + name.name();
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/entity/Role.java]:
```suggestion
        return name.name();
```

File: src/main/java/com/bookstore/entity/User.java [hunk 17]
```diff
@@ -5,30 +5,85 @@
...
-@Setter
-@Getter
 @Table(name = "users")
-public class User {
+@SQLDelete(sql = "UPDATE users SET is_deleted = true WHERE id=?")
+@SQLRestriction("is_deleted = false")
+@Getter
+@Setter
+@EqualsAndHashCode
+@ToString
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/entity/User.java]:
```suggestion
```

File: src/main/java/com/bookstore/entity/User.java [hunk 18]
```diff
@@ -5,30 +5,85 @@
...
     private Long id;
-
     @Column(nullable = false, unique = true)
     private String email;
-
     @Column(nullable = false)
     private String password;
-
-    @Column(nullable = false)
+    @Column(name = "first_name", nullable = false)
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/entity/User.java]:
```suggestion
    @Column(nullable = false)
```

File: src/main/java/com/bookstore/entity/User.java [hunk 19]
```diff
@@ -5,30 +5,85 @@
...
     private String firstName;
-
-    @Column(nullable = false)
+    @Column(name = "last_name", nullable = false)
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/entity/User.java]:
```suggestion
    @Column(nullable = false)
```

File: src/main/java/com/bookstore/entity/User.java [hunk 20]
```diff
@@ -5,30 +5,85 @@
...
     private String lastName;
-
+    @Column(name = "shipping_address")
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/entity/User.java]:
```suggestion
```

File: src/main/java/com/bookstore/entity/User.java [hunk 21]
```diff
@@ -5,30 +5,85 @@
...
     private String shippingAddress;
+
+    @ManyToMany
+    @JoinTable(
+            name = "users_roles",
+            joinColumns = @JoinColumn(name = "user_id"),
+            inverseJoinColumns = @JoinColumn(name = "role_id")
+    )
+    @ToString.Exclude
+    @EqualsAndHashCode.Exclude
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/entity/User.java]:
```suggestion
```

File: src/main/java/com/bookstore/entity/User.java [hunk 22]
```diff
@@ -5,30 +5,85 @@
...
+    private Set<Role> roles = new HashSet<>();
+
+    @Column(name = "is_deleted", nullable = false)
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/entity/User.java]:
```suggestion
    @Column(nullable = false)
```

File: src/main/java/com/bookstore/security/CustomUserDetailsService.java [hunk 23]
```diff
@@ -0,0 +1,22 @@
...
+
+@Service
+@RequiredArgsConstructor
+public class CustomUserDetailsService implements UserDetailsService {
+    private final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String email)
+            throws UsernameNotFoundException {
+        return userRepository.findByEmail(email)
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/security/CustomUserDetailsService.java:19]: Add to UserRepository
```
@EntityGraph(attributePaths = "roles")
Optional<User> findByEmail(String email);
```

File: src/main/resources/db/changelog/changes/02-create-users-table.yaml [hunk 24]
```diff
@@ -38,3 +38,23 @@ databaseChangeLog:
...
                   name: shipping_address
                   type: varchar(255)
+              - column:
+                  name: is_deleted
+                  type: boolean
+                  defaultValueBoolean: false
+                  constraints:
+                    nullable: false
+
+  - changeSet:
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/02-create-users-table.yaml]: remove changeSet

File: src/main/resources/db/changelog/changes/06-add-is-deleted-to-users.yaml [hunk 25]
```diff
@@ -0,0 +1,14 @@
+databaseChangeLog:
+  - changeSet:
+      id: add-is-deleted-to-users
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/06-add-is-deleted-to-users.yaml]: remove changeSet

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/com/bookstore/repository/RoleRepository.java [hunk 26]
```diff
@@ -0,0 +1,11 @@
+package com.bookstore.repository;
+
+import com.bookstore.entity.Role;
+import java.util.Optional;
+import org.springframework.data.jpa.repository.EntityGraph;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface RoleRepository extends JpaRepository<Role, Long> {
+    @EntityGraph(attributePaths = "roles")
```
Comment (liliia-ponomarenko) [inline src/main/java/com/bookstore/repository/RoleRepository.java]:
```suggestion
```

File: src/main/java/com/bookstore/security/CustomUserDetailsService.java [hunk 27]
```diff
@@ -0,0 +1,22 @@
...
+                .orElseThrow(() -> new EntityNotFoundException("Can`t find user by email"));
```
Comment (liliia-ponomarenko) [inline src/main/java/com/bookstore/security/CustomUserDetailsService.java]: add the email parameter to the message

File: src/main/java/com/bookstore/service/UserServiceImpl.java [hunk 28]
```diff
@@ -26,6 +31,9 @@ public UserResponseDto register(UserRegistrationRequestDto requestDto) {
         }
         User user = userMapper.toEntity(requestDto);
         user.setPassword(passwordEncoder.encode(requestDto.getPassword()));
+        Role role = roleRepository.findByName(Role.RoleName.ROLE_USER)
+                .orElseThrow(() -> new EntityNotFoundException("Role not found"));
```
Comment (liliia-ponomarenko) [inline src/main/java/com/bookstore/service/UserServiceImpl.java]: the same add parameter name

File: y [hunk 29]
```diff
@@ -0,0 +1,4 @@
+    Host git.slygods.com
+    HostName git.slygods.com
+    User git
+    IdentityFile ~/.ssh/id_rsa_slygods
```
Comment (liliia-ponomarenko) [inline y]: Why do you need this file?

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s fix a few mistakes ;)

File: y (hunk 29 above)
Comment (Roma1990Melnichuk) [reply y]: without this file the keys at work were constantly lost

Comment (ahoienko) [reply y]: What do you mean lost? Basically work should be saved automatically without having any additional files, maybe you should make some google research to identify why your work is lost.

In previous task you added this file as well to save work?

**Review by ahoienko — CHANGES_REQUESTED** [review]
Left one question

**Review by sarakhmen — APPROVED** [review]

**Pull request merged**